open_spiel/games/battleship.cc

```diff
@@ -0,0 +1,639 @@
+  // ships, or (iii) damaged and sank one of the opponent's ships.
+  //
+  // To be able to figure that out, we will keep track of the damage that each
+  // of the opponent's ship has received so far. The vector `ship_damage`
+  // contains and updates this information as each player's shot is processed in
+  // order. Position i corresponds to the damage that the opponent's ship in
+  // position i of bs_game->configuration.ships has suffered.
+  std::vector<int> ship_damage(conf.ships.size(), 0);
+  // Since in general we might have repeated shots, we cannot simply increase
+  // the ship damage every time a shot hits a ship. For that, we keep track of
+  // whether a cell was already hit in the past. We reuse the
+  // serialization/deserialization routines for shots to map from (r, c) to cell
+  // index r * board_width + c.
+  std::vector<bool> cell_hit(conf.board_width * conf.board_height, false);
+
+  std::string information_state;
+  for (const auto& move : moves_) {
+    if (absl::holds_alternative<ShipPlacement>(move.action)) {
+      // The player observed *their own* ship placements.
+      if (move.player == player) {
+        absl::StrAppend(&information_state, "/");
+        absl::StrAppend(&information_state,
+                        absl::get<ShipPlacement>(move.action).ToString());
+      }
+    } else {
+      const Shot& shot = absl::get<Shot>(move.action);
+
+      if (move.player != player) {
+        // If the shot came from the opponent, the player has seen it.
+        absl::StrAppend(&information_state, "/oppshot_", shot.ToString());
+      } else {
+        const int cell_index = SerializeShotAction_(shot);
+
+        char shot_outcome = 'W';  // For 'water'.
+        for (int ship_index = 0; ship_index < conf.ships.size(); ++ship_index) {
+          const Ship& ship = conf.ships.at(ship_index);
+
+          // SAFETY: the call to FindShipPlacement_ is safe, because if we are
+          // here it means that all ships have been placed.
+          const ShipPlacement ship_placement =
+              FindShipPlacement_(ship, opponent);
+
+          if (ship_placement.CoversCell(shot)) {
+            if (!cell_hit[cell_index]) {
+              // This is a new hit: we have to increas the ship damage and mark
+              // the cell as already hit.
+              ++ship_damage.at(ship_index);
+              cell_hit.at(cell_index) = true;
+            }
+            if (ship_damage.at(ship_index) == ship.length) {
+              shot_outcome = 'S';  // For 'sunk'.
+            } else {
+              shot_outcome = 'H';  // For 'hit' (but not sunk).
+            }
+          }
+        }
+
+        // Otherwise, the player knows they shot, but also knows whether the
+        // shot hit the water, hit a ship (but did not sink it), or sank a
+        // ship.
+        absl::StrAppend(&information_state, "/shot_", shot.ToString(), ":");
+
+        // XXX(gfarina): Apparently appending chars to a string is not a thing
```

Comment:
I think you can do information_state.push_back(shot_outcome) which I believe is faster / recommended.

Reply:
Ah wow that's even better, thanks :-)